docs/proposals/20230130-go-configuration.md

```diff
@@ -0,0 +1,189 @@
+
+## Pitfalls of Current Solution
+
+Currently, we primarily use [jsonnet](https://jsonnet.org/) in Observatorium, to compose a "meta-project" library that allows generating configuration for a multi-tenant scalable observability system.
+
+While jsonnet is a commonly used configuration templating language, there are certain issues with how jsonnet acts as a front-level configuration interface, which causes it to not be preferable to end-users.
+
+Templating via jsonnet, while achievable, often results in complex, unreadable code, and due to no doc generation and low IDE support becomes virtually undiscoverable. Codifying and abstracting operational practices and knowledge in jsonnet is difficult and time-consuming, due to the lack of type safety and testing practices. In certain cases, pre-existing upstreams also cannot be used and jsonnet templating needs to be written manually.
+
+Furthermore, for a project like Observatorium, this leads to a complex-to-adopt and messy installation model. An end-user is forced to go through one of the below,
+* Scale the steep learning curve of jsonnet initially (if needed), and understand complex Observatorium jsonnet library internals and abstractions (without documentation or discoverability) to be able to configure it exactly according to needs (sometimes even rewrite for customizability)
+* Copy over pre-generated YAML in observatorium/observatorium and change it manually
+* Reference generated YAML and rewrite it in templating the end-user is comfortable with
+
+## Goals
+
+Compose Observatorium configuration as a Go-based library using configuration-as-code concepts, that allows end-users to,
+* Easily discover configuration abstractions, without having to go through steep configuration templating language learning curves and having to explore templating code in-depth
+* Easily utilize pre-defined abstractions for their specific environments, and be able to override configuration at will, without hassle
+* Be able to rely on tested configuration generation, battle-tested versioning via Go modules and automatic documentation via [goDoc](https://blog.golang.org/godoc-documenting-go-code)
+* Be able to pull in, their own upstreams directly by just importing Go struct for their own configuration (sidecars/config files, etc)
+* Have a consistent installation and usage model for Observatorium, as a customizable simple static configuration generation solution, with sane defaults
+
+## Non-Goals
+
+* Deprecating/removing jsonnet entirely. This is only proposing Go-based configuration as an alternative method of usage, but does not take away from exisiting jsonnet templating in any way
+* Operator or operator generation. This proposal only aims for static configuration generation, and is not related to applying this configuration or automating it
+
+## Audience of this proposal
+
+Observatorium Devs and Users.
```

Comment:
Users are part of the audience but the proposal doesn't outline what's the user's point of view. For example, if I deploy Observatorium today using the jsonnet libs & apply some customizations there I only need to: create a jsonnetfile, add the dependency, run the jsonnet bundler tools (`jb`), and then use the `jsonnet` cli to render them. 

What would be the steps to reuse configuration with this Go system?

Reply:
I don't think the steps for jsonnet are quite clear, here. There are a few more steps to using jsonnet: create a jsonnetfile, add the dependency, run the jsonnet bundler tools (jb), understand how Observatorium's jsonnet lib works and is constructed (without docs/IDE support) and hunt through generated YAMLs to figure out how end state is affected, customize by writing complex jsonnet for diff envs to get to state that fits user deployment pipeline, use the jsonnet cli to render them to JSON and then use a tool like gojsontoyaml to convert to YAML.

The path for Go-based one is: 
Have Go tools installed, create a Go module and import obs/obs, check an example to figure out observatorium config abstractions or just go to pkg.go.dev for godoc or just click through + autocomplete in IDE to understand abstractions, use said abstractions and customize struct for diff envs to fit user deploy pipeline, `go run` to get YAML you need

In both cases above, we don't consider steps about learning Go or jsonnet.

Reply:
The point of my questions was more to think about these steps and outline them in the proposal, food for thought. Particularly I am missing the point of view of the user of Observatorium in there. 

i.e. why would I have to understand Observatorium's jsonnet lib? Aren't the default values good enough there? Why in the Go based config I can simply "check examples and figure out" and not in jsonnet? 

Better outlining these things will make the proposal more attractive, imho.

Reply:
I think this is covered by the "Pros" section of the proposal. All of those point out what can be done in Go and not jsonnet or which one is easier. If this feels inadequate, would adding a few examples of this help here?